Adding attributes that control the motion blur when rendering VDBs in Arnold

mattigruener opened the pull request:
I have added:
- velocity_scale
- velocity_fps

There are two more that Arnold provides:
- velocity_grids
- velocity_outlier_threshold

John, do you want me to add them all while I'm at it? For now I've added just the ones that we've deemed useful in an internal project. Not sure if we or someone else will ever need the other two. `velocity_grids` is only useful if you want to store motion vectors in a grid with a different name than the ones Arnold detects by default.

I'm copying descriptions from the [mtoa docs](https://support.solidangle.com/display/A5AFMUG/Volume):

**Velocity Grids**
Either 1 vector grid (eg. v or vel) or 3 float grids (eg. vel.x, vel.y, vel.z) representing the velocity field, to be used for motion blur. No motion blur will occur if an invalid combination of grids or if no grids are specified. Note that all velocity grids declared here are also available as channels in the shading context.

**Velocity Scale**
A scale factor for the velocity field. A value of 0 disables motion blur.

**Velocity FPS**
A scale factor for the velocity field. A value of 0 disables motion blur.

**Velocity Threshold**
Controls filtering of noisy velocities resulting in the faster rendering of motion blur from physics simulations. The default value 0.001 should have little to no visual impact. Setting it to zero disables filtering entirely.

andrewkaufman commented:
> Do we see the ArnoldVDB proxy continuing to have a place? Or do we think the native loading of VDBs via the SceneReader will replace it entirely? Personally I think I'd like to see the latter

I'm completely on board with the SceneReader approach, to the point of removing the ArnoldVDB proxy while we're busy breaking things anyways. IE has never used that proxy in production, and I think newcomers would find the SceneReader more intuitive in any case.

> we might have performance concerns.

We might, but I bet we could address those independently, via Encapsulate after load, or via those "format specific context variables that affects the SceneInterface" that we've been discussing.

> In the short term, we need to at least document the precedence of these overlapping settings.

I'd vote the short term to be removal of the ArnoldVDB proxy. If you're worried about some users, we could ping the dev list just to check if anyone is relying on it (and opposed to switching for some reason).

mattigruener commented on python/GafferArnoldUI/ArnoldAttributesUI.py:
```diff
@@ -572,6 +576,30 @@ def __volumeSummary( plug ) :
+
+			"description",
+			"""
+			Allows to scale the vector used in motion blur computation.
```
I'm not sure if it supports anything other than VDBs. Anyone has an idea what mtoa does to render Maya's fluids? Looking at the docs though, this is what some of the parameter descriptions say:

**Filename**
The location of the VDB file. 

**Grids**
A list of OpenVDB grids to read and make available as channels in the volume shading context. 

**Automatic Step Size**
Sets the minimum voxel size defined in the VDB as the step size.

So if they mention VDB explicitly in the docs, maybe we just do the same?

johnhaddon replied:
> So if they mention VDB explicitly in the docs, maybe we just do the same?

Sounds good to me.

mattigruener commented:
To add just one data point to the discussion about which approach would be the most common in the future: Yesterday, I've had an artist asking me about changing density values in Gaffer so that he could assign the same shader to a bunch of grids, but could still control the density per grid. If we want to support that kind of thing, there's no way around the new approach, I guess.

johnhaddon commented:
> I'm completely on board with the SceneReader approach, to the point of removing the ArnoldVDB proxy while we're busy breaking things anyways. IE has never used that proxy in production...

No, but we have used a similar proxy, right? Are we confident enough in the SceneReader approach that you're willing to ditch the IE proxy as quickly as you are the ArnoldVDB node?

>> we might have performance concerns.
> We might, but I bet we could address those independently,

Yeah, I buy that - it might be a little while till we get there, but I think you're right.

> I'd vote the short term to be removal of the ArnoldVDB proxy

I'd be happier to keep it as a parallel option for at least a little while. How about we hide it from the menu, and add a note in the node documentation that there are better options now?

andrewkaufman commented:
> we have used a similar proxy, right? Are we confident enough in the SceneReader approach that you're willing to ditch the IE proxy as quickly as you are the ArnoldVDB node?

I'm happy to ditch the IE proxy too anyways. I won't speak for the whole group on that though.

> I'd be happier to keep it as a parallel option for at least a little while. How about we hide it from the menu, and add a note in the node documentation that there are better options now?

Sounds good to me.

mattigruener commented:
You were right about the hashing. I've also added @donboie's new `VDBObject` in the switch.

The `grids` and `velocity_grids` are supported by the backend now. I have not added either to the `ArnoldAttributes` yet, because I wasn't sure if you wanted to use the "halfway-house option" for both or just `grids`. Either way, do we want to support this as a StringVectorData attribute like in the current implementation or would you rather have it as a space separated list (like on `ArnoldVDB`)? 
In the former case we might consider adding a StringVector entry in the "+" menu on `CustomAttributes`? Should we maybe add that regardless of what we do with these new attributes?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -771,6 +777,28 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
 
+			if( m_volumeGrids && m_volumeGrids->readable().size() )
+			{
+				const vector<std::string> &v = m_volumeGrids->readable();
```
Can we replace this with `IECoreArnold::ParameterAlgo::dataToArray()`?

johnhaddon commented on python/GafferArnoldUI/ArnoldAttributesUI.py:
```diff
@@ -572,6 +576,30 @@ def __volumeSummary( plug ) :
+
+			"description",
+			"""
+			Scale the vector used in vdb  motion blur computation.
```
Scale**s** the vector (extra s). The prevailing style for descriptions is to write them such that you could prefix them with "This plug" or "This plug is" and they would form a complete sentence. No need to expose users to the weird imperative tense we use in commits. While I'm nitpicking, there's an extra space after vdb, and VDB should be capitalised.

johnhaddon commented on python/GafferArnoldUI/ArnoldAttributesUI.py:
```diff
@@ -572,6 +576,30 @@ def __volumeSummary( plug ) :
+
+			"description",
+			"""
+			Set the frame rate used in vdb motion blur computation.
```
Set**s**, and VDB.

Do we have any more information about this attribute? When should one use this vs the scale for instance, and how does it work?

mattigruener replied:
I've sent an email to the solid angle support to get a little bit more information about this parameter. Ideally they'd sent us the formula used to compute the final vector for blurring... Will report back.

mattigruener replied:
Solidangle has told me now that they do this to determine if motion blur needs to be computed at all:

```
#define AI_EPSILON         (1.0e-4f)
if ( velocity_scale / velocity_fps < AI_EPSILON ) then motion blur is disabled.
```

I'm just guessing here, but I think they're assuming those motion vectors to be in `<your favorite spatial unit> / seconds` and the fps parameter is a scale to convert to frames, while the other parameter is to scale the spatial unit or to lookdev the blur? Does that make sense?

mattigruener commented:
Carsten thinks we should also expose the threshold parameter. Not having it might require post-processing the VDBs which is low and requires more disk space. He reckons we should expose it unless there's any objections.

johnhaddon commented:
> Carsten thinks we should also expose the threshold parameter.

Cool - no objections from me.

> I also noticed that we might want to update UserPlugs at some point so that it allows setting the geometric interpretation on vector data.

I'm pretty sure we discussed this previously and decided against it. The CompoundDataPlug is used in situations where the distinction between vector/point/normal is relevant (typically because it's generating Data to be passed to a renderer, as here). But the user plugs are not used in any scenario where the distinction is relevant, so we'd just be confusing users with multiple options which all effectively do the same thing...

mattigruener commented:
Added a commit to support `velocity_outlier_threshold`. Happy to squash it in once you're done reviewing, @johnhaddon.

mattigruener commented:
I'm assuming the failed tests are timeout issues and have nothing to do with my code.

andrewkaufman commented:
Restarted the stalled Linux test. The Mac tests here are also failing with the same thing I saw on #2438 

```
Traceback (most recent call last):
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/Gaffer/WeakMethod.py", line 67, in __call__
    return m( *args, **kwArgs )
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/PlugLayout.py", line 508, in __plugDirtied
    if not self.visible() or plug.direction() != plug.Direction.In :
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/Widget.py", line 255, in visible
    return self.__qtWidget.isVisibleTo( relativeTo )
RuntimeError: Internal C++ object (PySide2.QtWidgets.QWidget) already deleted.
Traceback (most recent call last):
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/Gaffer/WeakMethod.py", line 67, in __call__
    return m( *args, **kwArgs )
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/PlugLayout.py", line 508, in __plugDirtied
    if not self.visible() or plug.direction() != plug.Direction.In :
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/Widget.py", line 255, in visible
    return self.__qtWidget.isVisibleTo( relativeTo )
RuntimeError: Internal C++ object (PySide2.QtWidgets.QWidget) already deleted.
Traceback (most recent call last):
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/Gaffer/WeakMethod.py", line 67, in __call__
    return m( *args, **kwArgs )
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/PlugLayout.py", line 508, in __plugDirtied
    if not self.visible() or plug.direction() != plug.Direction.In :
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/Widget.py", line 255, in visible
    return self.__qtWidget.isVisibleTo( relativeTo )
RuntimeError: Internal C++ object (PySide2.QtWidgets.QWidget) already deleted.
Traceback (most recent call last):
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/Gaffer/WeakMethod.py", line 67, in __call__
    return m( *args, **kwArgs )
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/PlugLayout.py", line 508, in __plugDirtied
    if not self.visible() or plug.direction() != plug.Direction.In :
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/Widget.py", line 255, in visible
    return self.__qtWidget.isVisibleTo( relativeTo )
RuntimeError: Internal C++ object (PySide2.QtWidgets.QWidget) already deleted.
Traceback (most recent call last):
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/Gaffer/WeakMethod.py", line 67, in __call__
    return m( *args, **kwArgs )
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/PlugLayout.py", line 508, in __plugDirtied
    if not self.visible() or plug.direction() != plug.Direction.In :
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/Widget.py", line 255, in visible
    return self.__qtWidget.isVisibleTo( relativeTo )
RuntimeError: Internal C++ object (PySide2.QtWidgets.QWidget) already deleted.
libc++abi.dylib: terminating with uncaught exception of type IECore::Exception: Traceback (most recent call last):
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/Gaffer/WeakMethod.py", line 67, in __call__
    return m( *args, **kwArgs )
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/PathWidget.py", line 231, in __pathChanged
    self.setText( str( path ) )
  File "/Users/travis/build/GafferHQ/gaffer/install/gaffer-0.43.0.0-osx/python/GafferUI/TextWidget.py", line 64, in setText
    self._qtWidget().setText( text )
RuntimeError: Internal C++ object (_LineEdit) already deleted.
./config/travis/runGafferTest: line 3: 24369 Abort trap: 6           ./install/gaffer-*/bin/gaffer test $test_module
Gaffer Test Failed : GafferSceneUITest
```

mattigruener commented:
Hey John,

I added the hacky way of retrieving shutter information that I think I mentioned in an eMail. How much pain am I causing you with this? Also, I'm not quite sure how you'd want me to tackle respecting the `gaffer:deformationBlur` option. I don't think we have a case where an attribute needs to access globals yet? Do we need to change the API a little to allow option-dependant attributes?

Thanks,

Matti.

johnhaddon commented:
> I added the hacky way of retrieving shutter information that I think I mentioned in an eMail. How much pain am I causing you with this?

Having something like this as a stopgap measure is OK with me. It needs a unit test though. I'm actually surprised that it works, since as far as I can tell we don't set the Arnold camera until _after_ we've assigned the attributes.

> Also, I'm not quite sure how you'd want me to tackle respecting the gaffer:deformationBlur option.

Yeah, I don't have that fully worked out yet I'm afraid. Like I said in the email, my hope is we can make this all "just work" in `RendererAlgo::objectSamples()`. If we pass multiple samples to the backend for the `gaffer:deformationBlur == true` case, then we automatically get the sample times we need to set `motion_start` and `motion_end` in Arnold (no need to access the globals!). And then for `gaffer:deformationBlur == false` we'd send only one object sample to Arnold, and the backend would know that motion blur was off and could act accordingly.

The issue with this is that when deformation blur is on, but an object is static, we automatically throw away the extra samples, and pass just a single sample to the renderer. In this one case, passing VDObjects to Arnold, we happen to want to do the dumb thing and pass _all_ the samples. If we can somehow make an exception for that one case, it all just works.

> I don't think we have a case where an attribute needs to access globals yet? Do we need to change the API a little to allow option-dependant attributes?

No, I think attributes and options should remain orthogonal as far as possible : any mixing of them introduces complexity and makes IPR harder. It might be that we need some mixing for the Arnold backend, but in that case we should strive to hide the complexity there rather than expose it in the API.

The extra API I'm considering instead is a way of making queries to the renderer backend, so that the Arnold backend can say "actually I do always need all the samples for that type of object". I suspect there will be other queries we might want to make of a backend in the future, so adding a generic query mechanism seems pretty reasonable, even if our first use of it might be a little oddball.

3Delight recently added support for VDB volumes, and I'm waiting to get my hands on the docs for that to see if it informs our thinking at all.

mattigruener commented:
Hey John,

I've added the unittest (and included the attributes - not sure if you want to keep that). I've put a comment in there with a link to the bits of code that look like they control the order in which cameras and objects are being translated. Does that make sense?

Thanks.

johnhaddon commented on python/GafferArnoldTest/IECoreArnoldPreviewTest/RendererTest.py:
```diff
@@ -2058,6 +2058,65 @@ def testBackground( self ) :
+			"testCamera",
+			IECoreScene.Camera(
+				parameters = {
+					"shutter_start" : 10.75,
```
This isn't how shutters are [meant to be specified](https://github.com/GafferHQ/gaffer/blob/master/include/GafferScene/Private/IECoreScenePreview/Renderer.h#L238); we're just getting lucky because the IECoreArnold camera converter copies over any parameters it doesn't know about. See the GafferDelight tests for an [example of the right way](https://github.com/GafferHQ/gaffer/blob/master/python/GafferDelightTest/IECoreDelightPreviewTest/RendererTest.py#L413).

johnhaddon commented on python/GafferArnoldTest/IECoreArnoldPreviewTest/RendererTest.py:
```diff
@@ -2058,6 +2058,65 @@ def testBackground( self ) :
+		# We are doing the same when translating actual Gaffer scenes here:
+		# https://github.com/GafferHQ/gaffer/blob/master/src/GafferScene/Render.cpp#L254
+		r.camera(
+			"testCamera",
```
Although this is adding a camera, it's not actually declaring it as the active render camera using the ["camera" option](https://github.com/GafferHQ/gaffer/blob/master/include/GafferScene/Private/IECoreScenePreview/Renderer.h#L105). See [`testRenderRegion()`](https://github.com/GafferHQ/gaffer/blob/master/python/GafferArnoldTest/IECoreArnoldPreviewTest/RendererTest.py#L105) for an example of how to do that.

If you take a look at the ass file this test generates, you'll see that it contains an additional default camera which has been generated automatically due to the lack of the "camera" option. And crucially, it's this default camera which is the one that appears in the arnold options. And what's more, it has `shutter_start == shutter_end == 0`!

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -768,6 +785,49 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
+					float shutterStart = AiNodeGetFlt( arnoldCamera, g_shutterStartArnoldString );
+					float shutterEnd = AiNodeGetFlt( arnoldCamera, g_shutterEndArnoldString );
+
+					AiNodeSetFlt( node, g_motionStartArnoldString, shutterStart );
```
This is trying to set the motion_start and motion_end parameters on _any_ node, not just volume nodes.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -768,6 +785,49 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
+			if( m_volumeGrids && m_volumeGrids->readable().size() )
+			{
+				AtArray *array = ParameterAlgo::dataToArray( m_volumeGrids.get(), AI_TYPE_STRING );
+				AiNodeSetArray( node, g_volumeGridsArnoldString, array );
```
This isn't checking that the node is a volume node or has a volume_grids parameter.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1262,8 +1322,40 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
+					h.append( m_stepSize );
+					h.append( m_volumePadding );
+
+					if( m_volumeGrids )
```
This is duplicating the code from the case above. I'll make a suggestions to improve this and some other probems in my main review.

mattigruener commented:
Hey John,

I believe I've addressed the outstanding comments - please let me know if I've missed anything...

> Refactor m_volumeGrids, m_velocityGrids, m_velocityScale, m_velocityFPS and m_velocityOutlierThreshold into a new Volume struct, just as we've done for PolyMesh and Curves etc. 

That and the changes in `hashGeometryInternal()` and `applyGeometry()` are addressed in 8501c1a. I've also added the comment about us depending on Arnold's behaviour there.

The test has been changed in f8220be.

Does that look ok to you? If so, I'll rebase this into something cleaner tomorrow morning.

Cheerio!

Matti.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1241,6 +1199,90 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
+				AiNodeSetArray( node, g_velocityGridsArnoldString, array );
+			}
+
+			if( velocityScale != 1.0f && AiNodeEntryLookUpParameter( AiNodeGetNodeEntry( node ), g_velocityScaleArnoldString ) )
```
I think this `AiNodeLookUpParameter()` call is redundant now we know we're only operating on volume nodes, isn't it?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1241,6 +1199,90 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
+				}
+			}
+
+			if( velocityFPS != 24.0f && AiNodeEntryLookUpParameter( AiNodeGetNodeEntry( node ), g_velocityFPSArnoldString ) )
```
Again, I think we can omit the `AiNodeEntryLookUpParameter()` here. Also, do we need the `velocityFPS != 24` check? Seems like we'd be better to always set the parameter, just in case the Arnold default changes at any point. Same goes for the checks for velocityOutlierThreshold below.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1322,39 +1364,13 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
-
-						h.append( m_velocityScale );
-						h.append( m_velocityFPS );
-						h.append( m_velocityOutlierThreshold );
```
Do we not need an `m_volume.hash( h )` call here?

mattigruener replied:
I've addressed that in an extra commit -> a126a40. The actual issue was that I've missed applying those attributes to anything but @donboie's shiny new VDBObjects which I was too focused on, I guess. 
I'm not merging in this one just yet so you can have a look at it. 

Also, when I was playing with the ArnoldVDB node it got me thinking that with this PR we're breaking all Objects that come with those parameters already set, aren't we? We're basically resetting those parameters to the defaults if the respective attribute isn't set to a non-default value?
In the case of ArnoldVDB we could potentially switch that node to set the velocity related parameters as attributes instead, but maybe we need a more general solution in the backend?

johnhaddon replied:
> Also, when I was playing with the ArnoldVDB node it got me thinking that with this PR we're breaking all Objects that come with those parameters already set, aren't we? We're basically resetting those parameters to the defaults if the respective attribute isn't set to a non-default value?

Hmm. True. While we are planning to phase out ArnoldVDB in favour of Don's shiny VDBObjects, I don't think we should break any existing workflows just yet. How about storing the new attributes as `optional<Whatever>` and then only setting the Arnold parameters if the attribute itself was actually set in Gaffer?

mattigruener replied:
I've given that a try in 76ae41a. Not sure if you like the additional templated function, but it seemed to me like this is the cleanest way of doing it?

Also, I'd be curious to hear your opinion on 

`velocityScale.get() > 0`

versus

`velocityScale > 0.0f`

While the latter is fancy, the former is more explicit and maybe easier to understand? Maybe doesn't matter - I opted for the former.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1235,9 +1236,20 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
-			h.append( velocityScale );
-			h.append( velocityFPS );
-			h.append( velocityOutlierThreshold );
+			if( velocityScale )
```
There's a subtle bug here. I think we're unlikely to hit the conditions necessary to trigger it, but it's worth fixing, and worth knowing about in general given how much we rely on hashes everywhere. In the situations where only one of these three attributes is set, we will only append one value to the hash. In this case nothing in the hash distinguishes _which_ attribute was set, so we get hash collisions for the `velocityScale == X` and the `velocityFPS == X` cases. Does that make sense? The easiest solution is to always hash something for each attribute, maybe with an unconditional `h.append( velocityScale.get_value_or( default ) )`...

mattigruener replied:
Makes perfect sense, of course. I'll put up a fix later today! :)

mattigruener replied:
Thanks for catching that one, John. I've addressed it in f2ed774. If this is alright, I'll do another round of rebasing.

johnhaddon replied:
Great - looks good. Thanks Matti! Once you've rebased could you just go ahead and merge?
Cheers...
John

mattigruener commented:
Rebased. Will merge now. Thanks for your help with this, John.
